### PR TITLE
Introduction of namespaced features (see #1286)

### PR DESCRIPTION
**Pull request by djc**
I think this basically covers all of the plans from #1286, although it still needs a bunch of tests and documentation updates. Submitting this PR to get some early feedback on the direction.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! I'm have a bit of a hard time following this though, mind adding some comments on the various `bail!` statements in the updated feature resolution loop?

Additionally, to recap, the idea here is that packages with this flag *only* activate optional dependencies through `crate:foo`, right?

**Comment by djc** [comment]
@alexcrichton I've added a whole bunch of comments.

To recap: the idea here is that features live in a separate namespace from dependencies. As such, when specifying dependencies in a feature value list (in the features table), you have to prefix dependencies with `crate:`. However, even after this PR, you do not have to specify a feature explicitly if you just have an optional dependency: for optional dependencies where no feature is explicitly defined, an implicit feature will be created -- this should help avoid boilerplate. Then again, if you do specify a feature of the same name as an optional dependency, it *must* include the `crate:feature` dependency.

Hope that makes sense.

File: src/cargo/core/summary.rs [hunk 1]
```diff
@@ -131,55 +137,135 @@ impl PartialEq for Summary {
...
-                    dep_name
-                ),
-                (&Crate(ref dep), None) => bail!(
-                    "Feature `{}` includes `{}` which is neither \
-                     a dependency nor another feature",
-                    feature,
                     dep
                 ),
-                (&CrateFeature(_, _), Some(_)) | (&Feature(_), _) => {}
+                (&CrateFeature(_, _), _, _) => {}
```
Comment (alexcrichton) [inline src/cargo/core/summary.rs]: Could this explicitly list `(&CrateFeature(_, _), true, _)` for exhaustiveness?

Comment (djc) [reply src/cargo/core/summary.rs]: Fixed.

**Comment by djc** [comment]
Added a number of test cases that also did a good job flushing a number of logic issues out of the implementation. What is required in terms of documentation?

One thing to consider: should `namespaced-features` default to `true` for `Edition2018`?

**Comment by djc** [comment]
I've rebased on latest master and added some documentation to the unstable section.

File: src/cargo/core/summary.rs [hunk 2]
```diff
@@ -131,55 +137,168 @@ impl PartialEq for Summary {
...
-                    feature,
                     dep
                 ),
-                (&CrateFeature(_, _), Some(_)) | (&Feature(_), _) => {}
+                (&CrateFeature(_, _), true, _) => {}
             }
             values.push(val);
         }
+
+        if !dependency_found {
```
[the remark this answers is not in the document]
Comment (djc) [reply src/cargo/core/summary.rs]: I added a whole bunch of comments, does that help? It spans across the for-loop (which includes the match) as a performance optimization: this way, we don't have to loop over the feature values twice. That's entirely premature (in the sense of not having done any performance measurements), so if the comments don't help enough, I could instead rewrite to check the `values` list after the loop.

File: src/cargo/util/toml/mod.rs [hunk 3]
```diff
@@ -763,6 +764,7 @@ impl TomlManifest {
             deps,
             me.features.clone().unwrap_or_else(BTreeMap::new),
             project.links.clone(),
+            project.namespaced_features.unwrap_or(false),
```
Comment (alexcrichton) [inline src/cargo/util/toml/mod.rs]: Ah I just remembered, can this be gated behind a feature as well? (setting this key at all)

**Comment by alexcrichton** [comment]
Ok thanks for the new comments, they look great! I think this is good to go with a feature gate and fixed tests!

File: src/cargo/util/toml/mod.rs (hunk 3 above)
Comment (djc) [reply src/cargo/util/toml/mod.rs]: Sure - not sure what kind of gate you had in mind?

**Comment by djc** [comment]
Huh, those test failures got introduced by merging a conflict wrong. Fixed now, and rebased on current master. Just let me know what kind of feature gate you want and this can (finally!) be merged.

File: src/cargo/util/toml/mod.rs (hunk 3 above)
Comment (alexcrichton) [reply src/cargo/util/toml/mod.rs]: Oh something like `cargo-features = ["namespaced-features"]` in `Cargo.toml`, you can see some other examples of this via `src/cargo/core/features.rs` I think

Comment (alexcrichton) [reply src/cargo/util/toml/mod.rs]: For now yeah we'll want that (the Cargo feature basically unlocks the `namespaced-features` config value).

Long term we may not have both but for the near-term I think we'll want that.

**Comment by djc** [comment]
Rebased on master and added the feature gate. Hope we can merge this now!

**Comment by alexcrichton** [comment]
Hm looks like CI may be failing?

**Pull request merged**